Changes to build scripts to enable building dpctl with a custom DPC++.

diptorupd opened the pull request.

oleksandr-pavlyk commented on scripts/build_for_develop.sh:
```diff
@@ -1,54 +1,8 @@
-export CFLAGS=-fPIC
+export CODE_COVERAGE=ON
 python setup.py clean --all
 python setup.py build develop
```
I suppose `python setup.py develop` is enough, using `build` just makes it build it without installing, and the subsequent `develop` command builds it again (not sure why)

diptorupd replied:
Fixed.

oleksandr-pavlyk commented:
What about Windows? Should we have `dpctl-capi/dbg_build.bat` as well?

diptorupd commented:
> What about Windows? Should we have `dpctl-capi/dbg_build.bat` as well?

I added the script for windows. For some reason, I can build but not run the gtests on Windows. It seems to be related to the version of oneAPI I have on my laptop.

diptorupd commented:
@PokhodenkoSA how can the C API gtests be run on Windows? If I do `ninja check` I can successfully build the tests, but after that get an error. Most likely the issue is the the `DPCTLSyclInterface.dll` is not found. It will be good to apply a fix if it is a small one and have the C API tests run on Windows and CI.

PokhodenkoSA commented:
> have the C API tests run on Windows and CI.

On Linux too. I did not find gtest run on Linux CI.

PokhodenkoSA commented:
@diptorupd what is the status of the task?

diptorupd commented:
@oleksandr-pavlyk I cleaned this PR and removed most of the changes to the Python package build scripts (there are still few cosmetic ones). The changes here are only related to our CMake scripts to build `DPCTLSyclInterface` using custom dpcpp. Once this one is merged we can do the rest of the clean up in #426 and complete this round of refactoring.

oleksandr-pavlyk commented:
The cmake module finder for LLVMCov seem to have a problem:

```
-- Found Lcov: ~/.conda/envs/idp/bin/lcov
CMake Error at ~/.conda_local/envs/idp/share/cmake-3.19/Modules/FindPackageHandleStandardArgs.cmake:218 (message):
  Could NOT find LLVMCov (missing: LLVMCov_FOUND) (Required is at least
  version "11")
Call Stack (most recent call first):
  ~/.conda_local/envs/idp/share/cmake-3.19/Modules/FindPackageHandleStandardArgs.cmake:582 (_FPHSA_FAILURE_MESSAGE)
  cmake/modules/FindLLVMCov.cmake:65 (find_package_handle_standard_args)
  CMakeLists.txt:188 (find_package)


-- Configuring incomplete, errors occurred!
```

and the same time:

```
(idp) [07:37:34 nuc04 dpctl-capi]$ llvm-cov --version
LLVM (http://llvm.org/):
  LLVM version 12.0.0git
  Optimized build.
  Default target: x86_64-unknown-linux-gnu
  Host CPU: skylake
```
```
(idp) [07:37:40 nuc04 dpctl-capi]$ which llvm-cov
/localdisk/intel/oneapi/compiler/2021.2.0/linux/bin/llvm-cov
```

oleksandr-pavlyk commented on dpctl-capi/cmake/modules/FindLLVMCov.cmake:
```diff
@@ -0,0 +1,69 @@
+
+# Check if a specific version of llvm-cov is required.
+if(LLVMCov_FIND_VERSION AND (DEFINED LLVMCov_VERSION))
+    string(COMPARE EQUAL
```
```suggestion
    string(COMPARE LESS_EQUAL
```

This changes fixes the issue of not finding `llvm-cov` when its version is higher than 11.

oleksandr-pavlyk replied:
I pushed the change, since that worked locally.

diptorupd replied:
Thanks. Part of the issue is we are using the archaic cmake 3.10 as minimum version required. Prior to Cmake 3.19 there is no way to specify range of versions and only exact version is supported that is why I was using Equal. But, I guess your fix is fine here.

oleksandr-pavlyk commented:
Linux CI failed with

```
  -- Found IntelSycl: /opt/intel/oneapi/compiler/latest/linux (Required is at least version "2021.2.0")
08:12:24
  CMake Error at $PREFIX/share/cmake-3.14/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
08:12:24
    Could NOT find LevelZero (missing: LEVEL_ZERO_LIBRARY)
```

oleksandr-pavlyk commented:
The CMake runs fine now with nightly build of `intel/llvm@sycl` using CMake 3.19 after the change I pushed to `FindIntelSycl.cmake`

Using CMake 3.20 I run into trouble with finding `GTest::mock`. It expected `libgmock.so` in Python environment, which is not there. 

Building of `program_interface` generates 10 `` __SYCL2020_DEPRECATED`` warnings, but that is expected.

The linking phase fails with message:

```
[  3%] Linking CXX executable dpctl_c_api_tests
/usr/bin/ld: cannot find /localdisk/workspace/repos/llvm/build/lib/clang/13.0.0/lib/linux/libclang_rt.profile-x86_64.a: No such file or directory
clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
```
but it may have to do with my setup more than changes in this PR.

oleksandr-pavlyk commented:
> Linux CI failed with
> 
> ```
>   -- Found IntelSycl: /opt/intel/oneapi/compiler/latest/linux (Required is at least version "2021.2.0")
> 08:12:24
>   CMake Error at $PREFIX/share/cmake-3.14/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
> 08:12:24
>     Could NOT find LevelZero (missing: LEVEL_ZERO_LIBRARY)
> ```

@PokhodenkoSA Should these be libraries be expected in CI ? Should we be setting `L0_LIB_DIR` and `L0_INCLUDE_DIR` to help CMake find them in some non-standard location?

diptorupd commented:
> 
> 
> > Linux CI failed with
> > ```
> >   -- Found IntelSycl: /opt/intel/oneapi/compiler/latest/linux (Required is at least version "2021.2.0")
> > 08:12:24
> >   CMake Error at $PREFIX/share/cmake-3.14/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
> > 08:12:24
> >     Could NOT find LevelZero (missing: LEVEL_ZERO_LIBRARY)
> > ```
> 
> @PokhodenkoSA Should these be libraries be expected in CI ? Should we be setting `L0_LIB_DIR` and `L0_INCLUDE_DIR` to help CMake find them in some non-standard location?

I do not think that is the problem here. `ze_loader` is definitely installed on our CI system (see the build logs for the other PRs). It is some change I inadvertently introduced that is messing things up. I see from the logs that the C++ compiler selected  to build `libDPCTLSyclInterface` is GCC as opposed to dpcpp as in other builds. I will investigate further and fix it. But may be a while before I get to this as it.

diptorupd commented:
@oleksandr-pavlyk 

I added the changes we discussed on Friday. The new `GetLevelZeroHeaders.cmake` module clones the Level Zero repo and checks out the latest tag to get the headers in the build directory and builds `libDPCTLSyclInteface` with those headers. I think the changes will help us build on Windows with Level Zero program compilation support once I make few other changes.

Few more changes:

- Removed the automatic setting of `CMAKE_C_COMPILER` and `CMAKE_CXX_COMPILER` inside `FindIntelSycl.cmake`. It is considered an anti-pattern to do so.
- I added another cmake module `GetProjectVersion.cmake` that uses Git tags to figure out the version number. I was already doing something similar inside the docs sub-directory. Extracted the code into a reusable module.
- Added API version and `soname` to the generated `libDPCTLSyclInteface` shared library.

I will fix the Windows build with Level zero support and add a `pkg-config.in` and that will be the extent to which I want to take this PR.

diptorupd commented:
@oleksandr-pavlyk Now this needs your expertise :) The issue with the Linux CI is that post adding soname to `libDPCTLSyclInterface` things are breaking in the Cython extensions.

diptorupd commented:
@oleksandr-pavlyk Tremendous!!
